src/sync_device.rs

```diff
@@ -0,0 +1,707 @@
+    /// headphone jack is plugged in (used to disable external speakers).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// use evdev::{Device, SwitchType};
+    /// let device = Device::open("/dev/input/event0")?;
+    ///
+    /// // Does the device report a laptop lid switch?
+    /// let supported = device
+    ///     .supported_switches()
+    ///     .map_or(false, |axes| axes.contains(SwitchType::SW_LID));
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn supported_switches(&self) -> Option<AttributeSet<'_, SwitchType>> {
+        self.raw.supported_switches()
+    }
+
+    /// Returns a set of supported LEDs on the device.
+    ///
+    /// Most commonly these are state indicator lights for things like Scroll Lock, but they
+    /// can also be found in cameras and other devices.
+    pub fn supported_leds(&self) -> Option<AttributeSet<'_, LedType>> {
+        self.raw.supported_leds()
+    }
+
+    /// Returns a set of supported "miscellaneous" capabilities.
+    ///
+    /// Aside from vendor-specific key scancodes, most of these are uncommon.
+    pub fn misc_properties(&self) -> Option<AttributeSet<'_, MiscType>> {
+        self.raw.misc_properties()
+    }
+
+    // pub fn supported_repeats(&self) -> Option<Repeat> {
+    //     self.rep
+    // }
+
+    /// Returns the set of supported simple sounds supported by a device.
+    ///
+    /// You can use these to make really annoying beep sounds come from an internal self-test
+    /// speaker, for instance.
+    pub fn supported_sounds(&self) -> Option<AttributeSet<'_, SoundType>> {
+        self.raw.supported_sounds()
+    }
+
+    /// Fetches and returns events from the kernel ring buffer, doing synchronization on SYN_DROPPED.
+    ///
+    /// By default this will block until events are available. Typically, users will want to call
+    /// this in a tight loop within a thread.
+    /// Will insert "fake" events.
+    pub fn fetch_events(&mut self) -> io::Result<impl Iterator<Item = InputEvent> + '_> {
+        let block_dropped = std::mem::take(&mut self.block_dropped);
+        let sync = if block_dropped {
+            self.prev_state.clone_from(&self.state);
+            self.raw.sync_state(&mut self.state)?;
+            SyncState::Keys {
+                time: crate::systime_to_timeval(&std::time::SystemTime::now()),
+                start: Key::new(0),
+            }
+        } else {
+            SyncState::None
+        };
+
+        self.raw.fill_events()?;
+
+        Ok(FetchEventsSynced {
+            dev: self,
+            range: 0..0,
+            consumed_to: 0,
+            sync,
+        })
+    }
+
+    #[cfg(feature = "tokio")]
+    pub fn into_event_stream(self) -> io::Result<EventStream> {
+        EventStream::new(self)
+    }
+}
+
+impl AsRawFd for Device {
+    fn as_raw_fd(&self) -> RawFd {
+        self.raw.as_raw_fd()
+    }
+}
+
+macro_rules! try_compensate {
```

Comment:
Spicy macro. Can you add some comments or a rustdoc with what this typically generates?